### PR TITLE
[CodeCamp2023-504]Add a new script to support the WBF (Weighted Box Fusion) function in mmdetection.

### PR DESCRIPTION
**Pull request by Morty-Xu**
## Motivation

weighted boxes fusion (https://arxiv.org/abs/1910.13302) is a method for fusing predictions from different object detection models, which utilizes confidence scores of all proposed bounding boxes to construct averaged boxes. This PR aims to add weighted boxes fusion to mmdetection.

## Modification

- Add a .py file named wbf.py(mmdetection/mmdet/models/utils/wbf.py), modify mmdetection/mmdet/models/utils/__init__.py
- Add a .py file named demo_muti_model.py(mmdetection/demo/demo_muti_model.py)
- Add a .py file named fuse_result.py(mmdetection/tools/analysis_tools/fuse_result.py)

## Use cases (Optional)

- demo for fusing prediction results using wbf.

`python demo/demo_muti_model.py demo/demo.jpg ./configs/faster_rcnn/faster-rcnn_r50_fpn_2x_coco.py  \
./configs/yolo/yolov3_d53_8xb8-320-273e_coco.py --checkpoint \ ./checkpoints/faster_rcnn_r50_fpn_2x_coco_bbox_mAP-0.384_20200504_210434-a5d8aa15.pth \
    ./checkpoints/yolov3_d53_320_273e_coco-421362b6.pth --weights 1 2`

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMPreTrain.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by hhaAndroid** [comment]
@Morty-Xu Thank you. We can provide a Python script in the "demo" folder where users can input multiple configurations and weights to perform inference on images and folders. Additionally, we can provide a script in "tools/analysis_tools" to test the dataset. Relevant documentation will be added to explain these functionalities.

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @Morty-Xu,

We'd like to express our appreciation for your valuable contributions to the mmdetection. Your efforts have significantly aided in enhancing the project's quality.
It is our pleasure to invite you to join our community thorugh Discord_Special Interest Group (SIG) channel. This is a great place to share your experiences, discuss ideas, and connect with other like-minded people. To become a part of the SIG channel, send a message to the moderator, OpenMMLab, briefly introduce yourself and mention your open-source contributions in the #introductions channel. Our team will gladly facilitate your entry. We eagerly await your presence. Please follow this link to join us: ​https://discord.gg/UjgXkPWNqA.

If you're on WeChat, we'd also love for you to join our community there. Just add our assistant using the WeChat ID: openmmlabwx. When sending the friend request, remember to include the remark "mmsig + Github ID".

Thanks again for your awesome contribution, and we're excited to have you as part of our community!

File: docs/zh_cn/user_guides/useful_tools.md [hunk 1]
```diff
@@ -109,6 +109,79 @@ python tools/analysis_tools/analyze_results.py \
        --show-score-thr 0.3
 ```
 
+## 多模型检测结果融合
+
+`tools/analysis_tools/fuse_results.py` 可使用Weighted Boxes Fusion(WBF)方法将多个模型的检测结果进行融合。（当前仅支持COCO格式）
```
Comment (hhaAndroid) [inline docs/zh_cn/user_guides/useful_tools.md:114]:
```suggestion
`tools/analysis_tools/fuse_results.py` 可使用 Weighted Boxes Fusion(WBF) 方法将多个模型的检测结果进行融合。（当前仅支持 COCO 格式）
```

File: docs/zh_cn/user_guides/useful_tools.md [hunk 2]
```diff
@@ -109,6 +109,79 @@ python tools/analysis_tools/analyze_results.py \
...
+  - `avg`: 取平均值，默认为此选项。
+  - `max`: 取最大值。
+  - `box_and_model_avg`: box和模型尺度的加权平均值。
+  - `absent_model_aware_avg`: 考虑缺失模型的加权平均值。
+- `--eval-single`: 是否评估每个单一模型，默认值为`False`。
+- `--save-fusion-results`: 是否保存融合结果，默认值为`False`。
+- `--out-dir`: 融合结果保存的路径。
+
+**样例**:
+假设你已经通过 `tools/test.py` 得到了3个模型的 json 格式的结果文件，路径分别为 './model_1.json', './model_2.json', './model_3.json'，真实标注框的文件路径为'./annotation.json'。
```
Comment (hhaAndroid) [inline docs/zh_cn/user_guides/useful_tools.md:150]: 我觉得是不是直接你去 mmdet 里面选择 3 个直接可用的模型来作为案例更好？ 这样用户还可以复现。例如你可以选择 faster rcnn+ atss+ cascade rcnn 三个作为案例

File: docs/zh_cn/user_guides/useful_tools.md (hunk 1 above)
Comment (hhaAndroid) [reply docs/zh_cn/user_guides/useful_tools.md:114]: 中英文符号间需要空格

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**